From 70acdbadc27c99dc558bf99835ea6d4f115e8c28 Mon Sep 17 00:00:00 2001
From: Tzung-Bi Shih <tzungbi@chromium.org>
Date: Tue, 20 Dec 2022 14:51:59 +0800
Subject: [PATCH] CHROMIUM: scripts: add initial skeleton

This is a squash of the following commits:

Part of "CHROMIUM: config: add configuration files and infrastructure"
CHROMIUM: chromeos: Add script to update smatch error whitelist
CHROMIUM: chromeos: kernelconfig: Drop build/ dir
CHROMIUM: scripts: kernelconfig: Rewrite in python, parallelize
CHROMIUM: scripts: kernelconfig: Enter chroot automatically
CHROMIUM: scripts: kernelconfig: add 'checkconfig'
CHROMIUM: config: Update scripts to be kernel family aware
CHROMIUM: config: Yell if CHROMEOS_KERNEL_FAMILY isn't set
CHROMIUM: chromeos: Add scripts OWNERS
CHROMIUM: chromeos: kernelconfig: Fix sloppiness
CHROMIUM: chromeos: kernelconfig: Factor out per-arch builds
CHROMIUM: chromeos: kernelconfig: Factor out splitconfig
CHROMIUM: chromeos: kernelconfig: Dynamically determine ARCHES
CHROMIUM: chromeos: kernelconfig: Drop --family
CHROMIUM: chromeos: scripts: Redirect OWNERS to central repo
CHROMIUM: PRESUBMIT: Add check for tech debt
CHROMIUM: PRESUBMIT: Support multi-bug for UPSTREAM-TASK
CHROMIUM: Add presubmit script to validate commit subject lines
CHROMIUM: check_tags: Add Reland TAG
CHROMIUM: PRESUBMIT: tech_debt: Add missing tag
CHROMIUM: Add FIXUP tag

Signed-off-by: Tzung-Bi Shih <tzungbi@chromium.org>
---

Notes:
    b3d99dcb19aeebd75b0a1562b02fee570405f61c
    5d7414b833dcb4275e37608dc3015d5e9dae345e
    942e474f4a1ef138d23ad2dbadb2d470e50d747b
    364fca16bd1fc055ffec3ad97235cb7461f05eba
    8f6bf1a3aad839690eb9bbac43ebb6625c6d2151
    e7c28c8d9f2be10fa2115807199add29190a2a6e
    5e9fa622b9775843891d97df8ca7c68540b1cdf0
    33cbc0db46a1a86456e7d8c23893522a09547862
    18bb7e69dc477f78c0b6c89f44d37d1881592198
    a9c6d23c4f673d52f1da4e3e40945861be165e5c
    122ddeedb59a864583de8c6c4ff8603b370230aa
    346f80adfb517ad1c92a3af8164764fc781a5a34
    0dc53a0632015cb7b16e3f7183831a352561d485
    e8fb7a211f7a6eae98afedf97e2aac763cbc7e00
    82c286d081fafcbb632e1584f1755794bde8ee41
    b6352d1b254d14a34b81c74c95bd9ace4e0e7669
    2a2f450a50d4794b18511824d49df476357d9979
    3a372336da3eb24182434f86a4bf6e275312f990
    a72177889d7369981e1f57b8a954176b0f637701
    f80121b301a26d30065ef55cce0f19394f12a993
    5511534cd4ed275db76437d6570322bad75ef834

 PRESUBMIT.cfg                            |   7 +
 chromeos/scripts/OWNERS                  |   3 +
 chromeos/scripts/README                  |   2 +
 chromeos/scripts/check_tags.py           |  80 +++++
 chromeos/scripts/generate-its-script.sh  | 155 +++++++++
 chromeos/scripts/kernelconfig            | 398 +++++++++++++++++++++++
 chromeos/scripts/prepareconfig           |  28 ++
 chromeos/scripts/pylintrc                | 178 ++++++++++
 chromeos/scripts/tech_debt.py            |  74 +++++
 chromeos/scripts/update_smatch_whitelist |  91 ++++++
 10 files changed, 1016 insertions(+)
 create mode 100644 chromeos/scripts/OWNERS
 create mode 100644 chromeos/scripts/README
 create mode 100755 chromeos/scripts/check_tags.py
 create mode 100755 chromeos/scripts/generate-its-script.sh
 create mode 100755 chromeos/scripts/kernelconfig
 create mode 100755 chromeos/scripts/prepareconfig
 create mode 100644 chromeos/scripts/pylintrc
 create mode 100755 chromeos/scripts/tech_debt.py
 create mode 100755 chromeos/scripts/update_smatch_whitelist

diff --git a/PRESUBMIT.cfg b/PRESUBMIT.cfg
index 807aa17d4033..36a30a575881 100644
--- a/PRESUBMIT.cfg
+++ b/PRESUBMIT.cfg
@@ -1,3 +1,10 @@
+[Hook Scripts]
+# Only lint chromeos/ files.
+cros lint = cros lint "$(for f in ${PRESUBMIT_FILES}; do [[ $f == */chromeos/* ]] && echo $f; done)"
+# Check technical debt tags.
+technical debt = chromeos/scripts/tech_debt.py
+commit message = chromeos/scripts/check_tags.py
+
 [Hook Overrides]
 checkpatch_check: true
 aosp_license_check: false
diff --git a/chromeos/scripts/OWNERS b/chromeos/scripts/OWNERS
new file mode 100644
index 000000000000..99b7a190afba
--- /dev/null
+++ b/chromeos/scripts/OWNERS
@@ -0,0 +1,3 @@
+set noparent
+
+include chromiumos/owners:v1:/kernel/OWNERS.build
diff --git a/chromeos/scripts/README b/chromeos/scripts/README
new file mode 100644
index 000000000000..47d03b376c7f
--- /dev/null
+++ b/chromeos/scripts/README
@@ -0,0 +1,2 @@
+For details on how to use these scripts, please see:
+http://dev.chromium.org/chromium-os/how-tos-and-troubleshooting/kernel-configuration
diff --git a/chromeos/scripts/check_tags.py b/chromeos/scripts/check_tags.py
new file mode 100755
index 000000000000..e93a9a24db3b
--- /dev/null
+++ b/chromeos/scripts/check_tags.py
@@ -0,0 +1,80 @@
+#!/usr/bin/env python3
+# SPDX-License-Identifier: GPL-2.0
+"""Run presubmit checks for subject lines
+
+This script is intended to be used by `repo`. For more information, please see
+https://chromium.googlesource.com/chromiumos/repohooks/+/refs/heads/main/README.md.
+
+Enforces the following rules:
+  - Each uploaded patch must have valid subject line tag
+    - ChromeOS only commits must not have FROMLIST:, FROMGIT:,
+      or UPSTREAM: tags
+    - Patches with FROMLIST:, FROMGIT:, BACKPORT:, or UPSTREAM: tags
+      must not include ChromeOS only files
+"""
+
+import os
+import subprocess
+import sys
+
+
+TAG_MISSING_MSG = 'The subject of this patch does not include a valid tag'
+BAD_TAG_MSG = 'Bad subject tag for patch touching ChromeOS specific files'
+
+UPSTREAM_TAGS = ('FROMLIST:', 'UPSTREAM:', 'FROMGIT:', 'BACKPORT:')
+OTHER_TAGS = ('CHROMIUM:', 'WIP:', 'TEST:', 'TEST-ONLY:', 'Revert', 'Reland',
+	      'FIXUP:')
+
+CHROMEOS_FILES = ('chromeos/', 'OWNERS', 'PRESUBMIT.cfg', 'unblocked_terms.txt')
+
+
+def repo_path():
+    """Return path to repository"""
+    return subprocess.check_output(
+        ['git', 'rev-parse', '--show-toplevel'], encoding='utf-8').strip()
+
+
+def includes_chromium(files):
+    """Check if the CL modifies files that belong to ChromiumOS"""
+
+    gitroot=repo_path()
+
+    return any(os.path.relpath(f, gitroot).startswith(CHROMEOS_FILES)
+               for f in files)
+
+
+def check_commit(commit, files):
+    """Inspect commit messages to validate subject line."""
+    commit_message = subprocess.check_output(
+        ['git', 'log', '-n', '1', '--format=%B', commit],
+        encoding='utf-8',
+    )
+
+    if not commit_message.startswith(UPSTREAM_TAGS + OTHER_TAGS):
+        print(TAG_MISSING_MSG, file=sys.stderr)
+        return False
+
+    if commit_message.startswith(UPSTREAM_TAGS) and includes_chromium(files):
+        print(BAD_TAG_MSG, file=sys.stderr)
+        return False
+
+    return True
+
+
+def main():
+    """Main function."""
+
+    presubmit_files = os.environ.get('PRESUBMIT_FILES')
+    if presubmit_files is None:
+        sys.exit('Need a value for PRESUBMIT_FILES')
+
+    presubmit_commit = os.environ.get('PRESUBMIT_COMMIT')
+    if presubmit_commit is None:
+        sys.exit('Need a value for PRESUBMIT_COMMIT')
+
+    if not check_commit(presubmit_commit, presubmit_files.splitlines()):
+        sys.exit(1)
+
+
+if __name__ == '__main__':
+    main()
diff --git a/chromeos/scripts/generate-its-script.sh b/chromeos/scripts/generate-its-script.sh
new file mode 100755
index 000000000000..858587c7423d
--- /dev/null
+++ b/chromeos/scripts/generate-its-script.sh
@@ -0,0 +1,155 @@
+#!/bin/bash
+
+# Generate an its script used to build a u-boot fitImage kernel binary
+# that contains the kernel as well as device trees used when booting it
+# via mkimage or dtc.
+
+# Using mkimage will insert extra hashes that dtc will not add.
+
+kernel_arch=${ARCH}
+its_out=/dev/stdout
+dtb_dir=.
+compression="none"
+
+show_usage()
+{
+cat <<-EOF
+$0 [options] <image> <dtbs>
+	-a <arch>          CPU architecture
+	-c <compression>   Compression to use (none|lzma|lz4)
+	-d <dtb_dir>       directory where dtb files exist
+	-h                 shows this message
+	-o <its_script>    output its script pathname
+	-r <ramdisk_path>  ramdisk image to include
+
+$0 Image dts/*.dtb | dtc -I dts -O dtb -p 1024 > Image.fit
+EOF
+}
+
+die()
+{
+	exit 1
+}
+
+OPTIND=1
+while getopts "a:c:d:ho:r:" opt; do
+	case "$opt" in
+		a)
+			kernel_arch=${OPTARG}
+			;;
+		c)
+			compression=${OPTARG}
+			;;
+		d)
+			dtb_dir=${OPTARG}
+			;;
+		h)
+			show_usage
+			exit 0
+			;;
+		o)
+			its_out=${OPTARG}
+			;;
+		r)
+			ramdisk_path=${OPTARG}
+			;;
+	esac
+done
+
+shift $((OPTIND-1))
+
+if [ -z "${kernel_arch}" ]; then
+	>&2 echo "ERROR: CPU architecture not specified"
+	show_usage
+	die
+fi
+
+image_path=${1}
+shift
+
+if [[ "${compression}" == "lzma" ]]; then
+	lzma -9 -z -f -k "${image_path}" || die
+	image_path="${image_path}.lzma"
+elif [[ "${compression}" == "lz4" ]]; then
+	lz4 -20 -z -f "${image_path}" "${image_path}.lz4" || die
+	image_path="${image_path}.lz4"
+elif [[ "${compression}" != "none" ]]; then
+	>&2 echo "ERROR: compression method unknown"
+	show_usage
+	die
+fi
+
+cat > "${its_out}" <<-EOF || die
+/dts-v1/;
+
+/ {
+	description = "Chrome OS kernel image with one or more FDT blobs";
+	#address-cells = <1>;
+
+	images {
+		kernel@1 {
+			data = /incbin/("${image_path}");
+			type = "kernel_noload";
+			arch = "${kernel_arch}";
+			os = "linux";
+			compression = "${compression}";
+			load = <0>;
+			entry = <0>;
+		};
+EOF
+
+if [ -n "${ramdisk_path}" ]; then
+	cat >> "${its_out}" <<-EOF || die
+		ramdisk@1 {
+			data = /incbin/("${ramdisk_path}");
+			type = "ramdisk";
+			arch = "${kernel_arch}";
+			os = "linux";
+			compression = "gzip";
+			load = <0>;
+			entry = <0>;
+		};
+	EOF
+fi
+
+iter=1
+for dtb in "$@" ; do
+	cat >> "${its_out}" <<-EOF || die
+		fdt@${iter} {
+			description = "$(basename ${dtb})";
+			data = /incbin/("${dtb_dir}/${dtb}");
+			type = "flat_dt";
+			arch = "${kernel_arch}";
+			compression = "none";
+			hash@1 {
+				algo = "sha1";
+			};
+		};
+	EOF
+	((++iter))
+done
+
+cat <<-EOF >>"${its_out}"
+	};
+	configurations {
+		default = "conf@1";
+EOF
+
+for i in $(seq 1 $((iter-1))) ; do
+	cat >> "${its_out}" <<-EOF || die
+		conf@${i} {
+			kernel = "kernel@1";
+	EOF
+	if [ -n "${ramdisk_path}" ]; then
+		cat >> "${its_out}" <<-EOF || die
+			ramdisk = "ramdisk@1";
+		EOF
+	fi
+	cat >> "${its_out}" <<-EOF || die
+			fdt = "fdt@${i}";
+		};
+	EOF
+done
+
+echo "};" >> "${its_out}"
+echo "};" >> "${its_out}"
diff --git a/chromeos/scripts/kernelconfig b/chromeos/scripts/kernelconfig
new file mode 100755
index 000000000000..8ade2db05b8b
--- /dev/null
+++ b/chromeos/scripts/kernelconfig
@@ -0,0 +1,398 @@
+#!/usr/bin/env python3
+
+# Copyright 2021 The Chromium OS Authors. All rights reserved.
+# Use of this source code is governed by a BSD-style license that can be
+# found in the LICENSE file.
+
+# pylint: disable=missing-function-docstring
+
+"""Chrome OS kernel configuration tool
+
+Script to merge all configs and run 'make oldconfig' on it to wade out bad
+juju, then split the configs into distro-commmon and flavour-specific parts
+
+See this page for more details:
+http://dev.chromium.org/chromium-os/how-tos-and-troubleshooting/kernel-configuration
+"""
+
+import argparse
+import difflib
+import fnmatch
+import glob
+import logging
+import os
+import re
+import shutil
+import subprocess
+import sys
+import tempfile
+from typing import Dict, List
+
+MODES = ['oldconfig', 'olddefconfig', 'editconfig', 'genconfig', 'checkconfig']
+TOOLCHAIN_PREFIXES = {
+    'x86_64': 'x86_64-cros-linux-gnu',
+    'armel': 'armv7a-cros-linux-gnueabihf',
+    'arm64': 'aarch64-cros-linux-gnu',
+}
+
+
+def splitconfig(srcdir: str) -> Dict[str, str]:
+    """Split and merge common configs
+
+    Return dict of filename-to-contents.
+    """
+    allconfigs = {}
+
+    # Parse config files.
+    for config in os.listdir(srcdir):
+        # Only config.*
+        if not config.endswith('.config'):
+            continue
+
+        allconfigs[config] = set()
+
+        for line in open(os.path.join(srcdir, config)):
+            m = re.match(r'#*\s*CONFIG_(\w+)[\s=](.*)$', line)
+            if not m:
+                continue
+            option, value = m.groups()
+            allconfigs[config].add((option, value))
+
+    # Split out common config options.
+    common = None
+    for config in allconfigs:
+        if common is None:
+            common = allconfigs[config].copy()
+        else:
+            common &= allconfigs[config]
+    for config in allconfigs:
+        allconfigs[config] -= common
+    allconfigs['common.config'] = common
+
+    ret = {}
+    # Generate new splitconfigs.
+    for config in allconfigs:
+        if allconfigs[config] is None:
+            continue
+        contents = '#\n# Config options generated by splitconfig\n#\n'
+        for option, value in sorted(list(allconfigs[config])):
+            if value == 'is not set':
+                contents += '# CONFIG_%s %s\n' % (option, value)
+            else:
+                contents += 'CONFIG_%s=%s\n' % (option, value)
+        ret[config] = contents
+    return ret
+
+
+def die(msg: str = 'Aborting'):
+    logging.error(msg)
+    sys.exit(1)
+
+
+def in_chroot() -> bool:
+    return os.path.exists('/etc/cros_chroot_version')
+
+
+def rerun_in_chroot():
+    cmd = ['cros_sdk', '--working-dir', '.', '--',
+           os.path.relpath(sys.argv[0])] + sys.argv[1:]
+    logging.info('Re-running in chroot')
+    logging.debug('Run: %s', cmd)
+    sys.exit(subprocess.run(cmd, check=False).returncode)
+
+
+def make_toolchain_args(arch: str) -> Dict[str, str]:
+    ret = {
+        'LD': 'ld.lld',
+    }
+
+    if in_chroot():
+        prefix = TOOLCHAIN_PREFIXES[arch]
+        ccompiler = prefix + '-clang'
+        # CrOS build tooling wants us to use these, and avoid a default of
+        # unprefixed 'gcc' or 'clang'. Supply defaults here, but also consider
+        # environment varibles below.
+        ret['HOSTCC'] = 'x86_64-pc-linux-gnu-clang'
+        ret['HOSTCXX'] = 'x86_64-pc-linux-gnu-clang++'
+    else:
+        prefix = 'x86_64-linux-gnu'
+        ccompiler = prefix + '-gcc'
+
+    if not shutil.which(ccompiler):
+        if in_chroot():
+            die('%s not found. Try running "`sudo cros_setup_toolchains -t %s`"'
+                % (ccompiler, prefix))
+        else:
+            die('%s not found. Try running inside chroot.' % ccompiler)
+
+    # Propagate a few environment variables as make args.
+    for key in ('HOSTCC', 'HOSTCXX'):
+        val = os.getenv(key)
+        if val:
+            ret[key] = val
+
+    ret.update({
+        'CC': ccompiler,
+        'CXX': prefix + '-g++',
+        'CROSS_COMPILE': prefix + '-',
+    })
+    return ret
+
+
+def getchar() -> str:
+    # We don't want line-buffering, and shelling out to bash is much simpler
+    # than messing with tty settings.
+    return subprocess.check_output(
+            ['bash', '-c', 'read -n 1 -s; echo "${REPLY}"']
+            ).decode('utf-8').strip()
+
+
+def editconfig_prompt(variant: str) -> bool:
+    print('* %s: press <Enter> to edit, S to skip' % variant)
+    return getchar() not in ('S', 's')
+
+
+def filter_match(target: str, pattern: str) -> bool:
+    return fnmatch.fnmatch(target, '*' + pattern + '*')
+
+
+def is_editconfig_interactive(arch_flavour: str, prompt: bool,
+                              pattern: str) -> bool:
+    if filter_match(arch_flavour, pattern):
+        if not prompt:
+            return True
+        return editconfig_prompt(arch_flavour)
+    return False
+
+
+def make_cmd(builddir: str, arch: str, target: str) -> List[str]:
+    pairs = make_toolchain_args(arch)
+    pairs['ARCH'] = 'arm' if arch == 'armel' else arch
+    pairs['O'] = builddir
+
+    make_args = [k + '=' + v for k, v in pairs.items()]
+
+    return ['make', '-j'] + make_args + [target, 'savedefconfig']
+
+
+def checkconfig(srcdir: str, outdir: str) -> int:
+    error = 0
+    for src_config in glob.glob(srcdir + '/chromeos/config/**/*.config',
+                                recursive=True):
+        relative = os.path.relpath(src_config, srcdir)
+        out_config = os.path.join(outdir, relative)
+        logging.debug('Compare %s to %s', src_config, out_config)
+        before = open(src_config, 'r').readlines()
+        after = open(out_config, 'r').readlines()
+
+        if before == after:
+            continue
+
+        error = 1
+        diff = difflib.context_diff(before, after, fromfile='a/' + relative,
+                                    tofile='b/' + relative)
+        sys.stdout.writelines(diff)
+        logging.error('checkconfig failed for config %s', relative)
+
+    if error:
+        logging.error("""Consider running `%s olddefconfig` to normalize.
+
+NOTE: chromeos/config changes should always be in a CL by themselves and never
+squashed into the same patch as code changes. If code and config changes need
+to land together, consider using Cq-Depend to make a circular dependency.""",
+                      sys.argv[0])
+    else:
+        logging.info('All good!')
+    return error
+
+
+def build_one_arch(args, tmpdir: str, srcdir: str, family: str,
+                   arch: str) -> List[subprocess.Popen]:
+    procs = []
+    config_base_dir = srcdir + '/chromeos/config/' + family
+    config_arch_dir = config_base_dir + '/' + arch
+    for flavourconfig in glob.glob(os.path.join(srcdir, config_arch_dir,
+        '*.flavour.config')):
+        flavour = os.path.basename(flavourconfig)
+        builddir = os.path.join(tmpdir, 'build', family, arch, flavour)
+        os.makedirs(builddir)
+
+        # Merge base.config, common.config, <flavour>.flavour.config
+        conf = '\n'.join(open(os.path.join(srcdir, f)).read() for f in [
+                config_base_dir + '/base.config',
+                config_arch_dir + '/common.config',
+                config_arch_dir + '/' + flavour,
+        ])
+        open(os.path.join(builddir, '.config'), 'w').write(conf)
+
+        interactive = False
+        if args.mode == 'genconfig':
+            mode = 'olddefconfig'
+        elif args.mode == 'editconfig':
+            interactive = is_editconfig_interactive(
+                    family + '/' + arch + '/' + flavour,
+                    not args.yes, args.filter)
+            mode = 'menuconfig' if interactive else 'olddefconfig'
+        elif args.mode == 'checkconfig':
+            mode = 'olddefconfig'
+        else:
+            if args.mode == 'oldconfig':
+                # 'oldconfig' may run into interactive prompts.
+                interactive = True
+            mode = args.mode
+
+        cmd = make_cmd(builddir, arch, mode)
+        if interactive:
+            logging.info('Starting interactive cmd: %s', ' '.join(cmd))
+            ret = subprocess.run(cmd, check=False, cwd=srcdir)
+            if ret.returncode:
+                die('cmd "%s" failed' % ' '.join(cmd))
+        else:
+            logging.info('Starting background cmd: %s', ' '.join(cmd))
+            procs += [subprocess.Popen(cmd, cwd=srcdir,
+                                        stdout=subprocess.PIPE,
+                                        stderr=subprocess.PIPE)]
+    return procs
+
+
+def do_splitconfig(tmpdir: str, srcdir: str, outdir: str,
+                   family: str, arches: List[str], save_configs: bool):
+    family_cfg_dir = 'chromeos/config/' + family
+
+    if save_configs:
+        keep_dir = 'CONFIGS/' + family
+        os.makedirs(keep_dir, exist_ok=True)
+
+    for arch in arches:
+        dest = os.path.join(tmpdir, family, arch)
+        family_arch_dir = family_cfg_dir + '/' + arch
+        os.makedirs(dest)
+        for flavourconfig in glob.glob(os.path.join(srcdir, family_arch_dir,
+                                       '*.flavour.config')):
+            flavour = os.path.basename(flavourconfig)
+            builddir = os.path.join(tmpdir, 'build', family, arch, flavour)
+            shutil.copy(os.path.join(builddir, 'defconfig'),
+                        os.path.join(dest, flavour))
+            if save_configs:
+                shutil.copy(os.path.join(builddir, '.config'),
+                            os.path.join(keep_dir, arch + '-' + flavour))
+                shutil.copy(os.path.join(builddir, 'defconfig'),
+                            os.path.join(keep_dir, arch + '-' +
+                                         flavour + '.def'))
+        os.makedirs(os.path.join(outdir, family_arch_dir), exist_ok=True)
+        # Find per-arch common items; flavour-unique options can be written out
+        # immediately.
+        for config, contents in splitconfig(dest).items():
+            if config == 'common.config':
+                open(os.path.join(tmpdir, family, arch + '.config'),
+                     'w').write(contents)
+            else:
+                open(os.path.join(outdir, family_arch_dir, config),
+                     'w').write(contents)
+
+    # Find cross-arch common items; common ones go to base.config, and unique
+    # ones to <arch>/common.config.
+    for config, contents in splitconfig(tmpdir + '/' + family).items():
+        if config == 'common.config':
+            open(os.path.join(outdir, family_cfg_dir, 'base.config'),
+                 'w').write(contents)
+        else:
+            assert config.endswith('.config')
+            # Python3.9: arch = config.removesuffix('.config')
+            arch = config[:-len('.config')]
+            open(os.path.join(outdir, family_cfg_dir, arch, 'common.config'),
+                 'w').write(contents)
+
+
+# Arches relevant to this family.
+def family_arches(family: str) -> List[str]:
+    return [os.path.basename(p)
+            for p in glob.glob('chromeos/config/' + family + '/*')
+            if os.path.isdir(p)]
+
+
+def doit(args, tmpdir: str) -> int:
+    for d in (os.getcwd(),
+              os.path.abspath(os.path.join(
+                  os.path.dirname(sys.argv[0]), '..', '..'))):
+        if (os.path.exists(os.path.join(d, 'MAINTAINERS')) and
+            os.path.exists(os.path.join(d, 'Makefile'))):
+            srcdir = d
+            logging.info('Using top kernel dir: %s', srcdir)
+            break
+    else:
+        die('This does not appear to be a kernel source directory.')
+
+    if not in_chroot() and not args.force:
+        rerun_in_chroot()
+        assert False # Should not reach this.
+
+    # Most commands write back to source directory.
+    if args.mode == 'checkconfig':
+        outdir = tmpdir
+    else:
+        outdir = srcdir
+
+    save_configs = args.mode == 'genconfig'
+
+    procs = []
+
+    families = [os.path.basename(p)
+                for p in glob.glob('chromeos/config/*')
+                if os.path.isdir(p)]
+    for family in families:
+        for arch in family_arches(family):
+            if not arch in TOOLCHAIN_PREFIXES:
+                die('Unexpected arch: %s' % arch)
+            procs += build_one_arch(args, tmpdir, srcdir, family, arch)
+
+    for p in procs:
+        ret = p.wait()
+        if ret:
+            assert p.stderr # Remind pytype we captured stderr.
+            logging.error(p.stderr.read().decode('utf-8'))
+            die('cmd "%s" failed' % ' '.join(p.args))
+
+    logging.info('Generating splitconfigs')
+    for family in families:
+        do_splitconfig(tmpdir, srcdir, outdir,
+                       family, family_arches(family), save_configs)
+
+    if args.mode == 'checkconfig':
+        return checkconfig(srcdir, outdir)
+
+    return 0
+
+
+def main() -> int:
+    parser = argparse.ArgumentParser(
+            description='Chrome OS kernel configuration script',
+            epilog="""
+               Note that Kbuild will evaluate some features depending on the
+               toolchain, so we try to enter the SDK chroot. This can be
+               overridden, with a potentially degraded experience.
+               """)
+
+    log_level_choices = ['DEBUG', 'INFO', 'WARNING', 'ERROR', 'CRITICAL']
+    parser.add_argument(
+        '--log_level', '-l',
+        choices=log_level_choices,
+        default='INFO'
+    )
+    parser.add_argument('--force', '-F', action='store_true',
+                        help='Force allowing to run outside the chroot')
+    parser.add_argument('--filter', '-f', default='',
+                        help='Only attempt to edit configs which match filter')
+    parser.add_argument('--yes', '-y', action='store_true',
+                        help='Edit all configs which match unconditionally')
+    parser.add_argument('mode', choices=MODES, help='sub-command/mode')
+
+    args = parser.parse_args()
+    logging.basicConfig(level=args.log_level,
+                        format='%(levelname)s - %(message)s')
+
+    with tempfile.TemporaryDirectory() as tmpdir:
+        return doit(args, tmpdir)
+
+
+sys.exit(main())
diff --git a/chromeos/scripts/prepareconfig b/chromeos/scripts/prepareconfig
new file mode 100755
index 000000000000..889f4f40a610
--- /dev/null
+++ b/chromeos/scripts/prepareconfig
@@ -0,0 +1,28 @@
+#!/bin/bash
+
+# See this page for more details:
+# http://dev.chromium.org/chromium-os/how-tos-and-troubleshooting/kernel-configuration
+
+if [[ -z "${CHROMEOS_KERNEL_FAMILY}" ]]; then
+    echo "CHROMEOS_KERNEL_FAMILY env variable not set; see crrev.com/c/3398813" 1>&2
+    echo "If running manually, try CHROMEOS_KERNEL_FAMILY=chromeos" 1>&2
+    exit 1
+fi
+
+family="${CHROMEOS_KERNEL_FAMILY}"
+
+flavourconf=$(find chromeos/config/${family} -name $1.flavour.config)
+if [ ! -f "${flavourconf}" ]; then
+    echo "Found no flavour configuration for '$1'." 1>&2
+    exit 1
+fi
+
+outputfile="${2:-.config}"
+
+archconfdir=$(dirname ${flavourconf})
+arch=$(basename ${archconfdir})
+
+# Generate .config
+cat "chromeos/config/${family}/base.config" \
+    "${archconfdir}/common.config" \
+    "${flavourconf}" > "${outputfile}"
diff --git a/chromeos/scripts/pylintrc b/chromeos/scripts/pylintrc
new file mode 100644
index 000000000000..2d09c22ceb2b
--- /dev/null
+++ b/chromeos/scripts/pylintrc
@@ -0,0 +1,178 @@
+# Copyright 2017 The Chromium OS Authors. All rights reserved.
+# Use of this source code is governed by a BSD-style license that can be
+# found in the LICENSE file.
+
+[MASTER]
+
+load-plugins=chromite.cli.cros.lint
+
+# Configure quote preferences.
+string-quote = single-avoid-escape
+triple-quote = double
+docstring-quote = double
+
+[MESSAGES CONTROL]
+
+enable=
+    apply-builtin,
+    backtick,
+    bad-python3-import,
+    buffer-builtin,
+    cmp-builtin,
+    cmp-method,
+    coerce-builtin,
+    coerce-method,
+    delslice-method,
+    deprecated-itertools-function,
+    deprecated-str-translate-call,
+    deprecated-string-function,
+    deprecated-types-field,
+    dict-items-not-iterating,
+    dict-iter-method,
+    dict-keys-not-iterating,
+    dict-values-not-iterating,
+    dict-view-method,
+    div-method,
+    exception-message-attribute,
+    execfile-builtin,
+    file-builtin,
+    filter-builtin-not-iterating,
+    getslice-method,
+    hex-method,
+    idiv-method,
+    import-star-module-level,
+    indexing-exception,
+    input-builtin,
+    intern-builtin,
+    invalid-str-codec,
+    long-builtin,
+    map-builtin-not-iterating,
+    metaclass-assignment,
+    next-method-called,
+    next-method-defined,
+    nonzero-method,
+    oct-method,
+    old-raise-syntax,
+    parameter-unpacking,
+    print-statement,
+    raising-string,
+    range-builtin-not-iterating,
+    raw_input-builtin,
+    rdiv-method,
+    reduce-builtin,
+    reload-builtin,
+    setslice-method,
+    standarderror-builtin,
+    sys-max-int,
+    unichr-builtin,
+    unpacking-in-except,
+    using-cmp-argument,
+    xrange-builtin,
+    zip-builtin-not-iterating,
+
+disable=
+    too-many-lines,
+    too-many-branches,
+    too-many-statements,
+    too-few-public-methods,
+    too-many-instance-attributes,
+    too-many-public-methods,
+    too-many-locals,
+    too-many-arguments,
+    locally-enabled,
+    locally-disabled,
+    fixme,
+    bad-continuation,
+    invalid-name,
+
+[REPORTS]
+
+# Tells whether to display a full report or only the messages
+# CHANGE: No report.
+reports=no
+
+# Activate the evaluation score.
+score=no
+
+
+[FORMAT]
+
+# Maximum number of characters on a single line.
+max-line-length=80
+
+
+[TYPECHECK]
+
+# List of members which are set dynamically and missed by pylint inference
+# system, and so shouldn't trigger E0201 when accessed.
+# CHANGE: Added 'AndReturn', 'InAnyOrder' and 'MultipleTimes' for pymox.
+# CHANGE: Added tempdir for @osutils.TempDirDecorator.
+generated-members=REQUEST,acl_users,aq_parent,AndReturn,InAnyOrder,MultipleTimes,tempdir
+
+
+[BASIC]
+
+# List of builtins function names that should not be used, separated by a comma.
+# exit & quit are for the interactive interpreter shell only.
+# https://docs.python.org/3/library/constants.html#constants-added-by-the-site-module
+bad-functions=
+    apply,
+    exit,
+    filter,
+    input,
+    map,
+    quit,
+    raw_input,
+    reduce,
+
+# Regular expression which should only match correct function names
+#
+# CHANGE: The ChromiumOS standard is different than PEP-8, so we need to
+# redefine this.
+#
+# Common exceptions to ChromiumOS standard:
+# - main: Standard for main function
+function-rgx=([A-Z_][a-zA-Z0-9]{2,30}|main)$
+
+# Regular expression which should only match correct method names
+#
+# CHANGE: The ChromiumOS standard is different than PEP-8, so we need to
+# redefine this. Here's what we allow:
+# - CamelCaps, starting with a capital letter.  No underscores in function
+#   names.  Can also have a "_" prefix (private method) or a "test" prefix
+#   (unit test).
+# - Methods that look like __xyz__, which are used to do things like
+#   __init__, __del__, etc.
+# - setUp, tearDown: For unit tests.
+method-rgx=((_|test)?[A-Z][a-zA-Z0-9]{2,30}|__[a-z]+__|setUp|tearDown)$
+
+
+[SIMILARITIES]
+
+# Minimum lines number of a similarity.
+min-similarity-lines=8
+
+
+[IMPORTS]
+
+# Deprecated modules which should not be used, separated by a comma.
+# Bastion: Dropped in Python 3.
+# mox: Use the 'mock' module instead.
+# optparse: Use the 'argparse' module instead.
+# regsub: Use the 're' module instead.
+# rexec: Dropped in Python 3.
+# TERMIOS: Use the 'termios' module instead.
+deprecated-modules=
+    Bastion,
+    mox,
+    optparse,
+    regsub,
+    rexec,
+    TERMIOS,
+
+
+[LOGGING]
+
+# Apply logging string format checks to calls on these modules.
+logging-modules=
+    logging,
diff --git a/chromeos/scripts/tech_debt.py b/chromeos/scripts/tech_debt.py
new file mode 100755
index 000000000000..8f773fdd622b
--- /dev/null
+++ b/chromeos/scripts/tech_debt.py
@@ -0,0 +1,74 @@
+#!/usr/bin/env python3
+"""Does presubmit checks for kernel technical debt.
+
+This is intended to be used by `repo`. For more information, please see
+https://chromium.googlesource.com/chromiumos/repohooks/+/refs/heads/main/README.md.
+
+Enforces the following rules:
+  - If a patch is not from upstream, a UPSTREAM-TASK tag must be present
+"""
+
+import os
+import re
+import subprocess
+import sys
+
+
+TECH_DEBT_MSG = 'This patch is not fully upstream. Please open a tracking bug here: go/cros-kernel-technical-debt-bug and add a label UPSTREAM-TASK=b:XXXX referencing it. A member of the review committee will review the CL. Thank you.'
+
+
+def is_chromium_only(files):
+    """Check if the CL modifies files that only belong to ChromiumOS"""
+    repo_path = subprocess.check_output(
+        ['git', 'rev-parse', '--show-toplevel'], encoding='utf-8').strip()
+
+    for file in files:
+        file = os.path.relpath(file, repo_path)
+        if not file.startswith(('chromeos/', 'OWNERS', 'PRESUBMIT.cfg',
+                                'unblocked_terms.txt')):
+            return False
+
+    return True
+
+
+def check_tech_debt(commit):
+    """Inspect commit messages to validate tech_debt tags."""
+    commit_message = subprocess.check_output(
+        ['git', 'log', '-n', '1', '--format=%B', commit],
+        encoding='utf-8',
+    )
+    if commit_message.startswith(
+        ('UPSTREAM:', 'FROMGIT:', 'WIP:', 'TEST:', 'TEST-ONLY:', 'BACKPORT:',
+         'Revert', 'Reland')) and not commit_message.startswith(
+        'BACKPORT: FROMLIST'):
+        return True
+
+    buganizer = r'b:[0-9]{7,}'
+    tag_line = r'\nUPSTREAM-TASK=%s([, ]+%s)*[ \t]*\n' % (buganizer, buganizer)
+    if re.search(tag_line, commit_message):
+        return True
+
+    print(TECH_DEBT_MSG, file=sys.stderr)
+    return False
+
+
+def main():
+    """Main function."""
+
+    presubmit_files = os.environ.get('PRESUBMIT_FILES')
+    if presubmit_files is None:
+        sys.exit('Need a value for PRESUBMIT_FILES')
+
+    presubmit_commit = os.environ.get('PRESUBMIT_COMMIT')
+    if presubmit_commit is None:
+        sys.exit('Need a value for PRESUBMIT_COMMIT')
+
+    if is_chromium_only(presubmit_files.splitlines()):
+        sys.exit(0)
+
+    if not check_tech_debt(presubmit_commit):
+        sys.exit(1)
+
+
+if __name__ == '__main__':
+    main()
diff --git a/chromeos/scripts/update_smatch_whitelist b/chromeos/scripts/update_smatch_whitelist
new file mode 100755
index 000000000000..6ebd458efa8c
--- /dev/null
+++ b/chromeos/scripts/update_smatch_whitelist
@@ -0,0 +1,91 @@
+#! /bin/bash
+#
+# Copyright (c) 2013 The Chromium OS Authors. All rights reserved.
+# Distributed under the terms of the GNU General Public License v2
+
+# Print help if no args provided, or if the user is requesting help.
+if [[ $# -eq 0 || $1 == "--help" || $1 == "-h" ]]; then
+	name=$(basename $0)
+	cat <<-EOF
+	USAGE: ./${name} [list of boards]
+	  e.g. ./${name} x86-generic amd64-generic daisy
+	  This command will build kernel for the following boards:
+	    - x86-generic
+	    - amd64-generic
+	    - daisy
+	  It will then generate a file containing all smatch errors not already
+	  in the known errors whitelist.  Duplicates will be listed only once.
+
+	  If chromeos-kernel is not cros_workon'd for a board, that board will
+	  be skipped.
+	EOF
+	exit
+fi
+
+# Put board name arguments into a list.
+BOARDS=()
+SKIPPED_BOARDS=()
+while [[ $# -gt 0 ]]; do
+	board=$1
+	# If kernel is not cros-workon'd, then skip this board.
+	if cros_workon-${board} list | grep -q chromeos-kernel; then
+		BOARDS+=( "${board}" )
+	else
+		SKIPPED_BOARDS+=( "${board}" )
+	fi
+	shift
+done
+
+datestamp() {
+	date +%Y%m%d%H%M%s
+}
+
+# We want to ignore smatch test errors when building kernel, so that we can
+# collect those errors for updating the whitelist.
+export FEATURES='test test-fail-continue'
+export USE="smatch ignore_test_errors"
+LOG_FILES=()
+for board in "${BOARDS[@]}"; do
+	LOG_FILE=/tmp/${board}-kernel-$(datestamp).log
+
+	emerge-${board} chromeos-kernel |& tee "${LOG_FILE}"
+	pipestatus=${PIPESTATUS[*]}
+	if [[ ${pipestatus// } -ne 0 ]]; then
+		echo "Error building kernel for ${board}" >&2
+		exit 1
+	fi
+	grep "Non-whitelisted error found:" "${LOG_FILE}" | cut -f 2- -d: | \
+		sed -e 's/"$//g' | cut -f 2- -d '"' | \
+		sort > "${LOG_FILE}.new_errors"
+	LOG_FILES+=( "${LOG_FILE}.new_errors" )
+done
+
+# Create a new log file to contain all errors, overwriting it if it already
+# exists.  For each board's new errors, add to this log file the errors that
+# haven't already been added.  This creates the union of all new errors.
+NEW_ERRORS_FILE=/tmp/smatch_new_errors-$(datestamp).log
+for log_file in "${LOG_FILES[@]}"; do
+	while read line; do
+		fgrep -q "${line}" "${NEW_ERRORS_FILE}" && continue
+		echo "${line}"
+	done < "${log_file}"
+done > "${NEW_ERRORS_FILE}"
+
+# Rename the new errors file to reflect the new boards, and move it to the
+# current working directory.
+BOARDS_NO_SPACES=$(printf -- -%s "${BOARDS[@]}")
+NEW_ERRORS_FILE_FINAL=smatch_new_errors${BOARDS_NO_SPACES}.log
+if [[ $(wc -l < "${NEW_ERRORS_FILE}") -ne 0 ]]; then
+	mv "${NEW_ERRORS_FILE}" "${NEW_ERRORS_FILE_FINAL}" || exit
+fi
+
+echo "*****************************************************"
+echo "Done building kernel for these boards: ${BOARDS[*]}"
+[[ ${#SKIPPED_BOARDS[@]} -gt 0 ]] && \
+	echo "Skipped these boards because kernel was not cros_workon'd:" \
+		"${SKIPPED_BOARDS[*]}"
+if [[ -e ${NEW_ERRORS_FILE_FINAL} ]]; then
+	echo "New smatch errors are in ${NEW_ERRORS_FILE_FINAL}"
+else
+	echo "No new smatch errors found."
+fi
-- 
2.39.0.314.g84b9a713c41-goog

